Add CI/CD workflows for terraform automation

DaMandal0rian opened the pull request:
The PR adds terraform automation with github actions CI for devnet, gemini-3(x) and ephemeral devnets.

Closes #184

DaMandal0rian committed:
Add CI/CD workflows for terraform automation

- add transcrypt script
- workflow changes to terraform

DaMandal0rian committed:
Permission fixes and paths

- fix permissions
- write to /tmp path and not directly in repo for transcrypt
- fix backend config path

nazar-pc commented on .github/workflows/devnet_main_deploy.yml:
```diff
@@ -0,0 +1,27 @@
+      - main
+  push:
+    paths:
+      - "aws/devnet/**"
```
So it will redeploy regardless of things changing if merged to `main`, which is probably fine, but also why?

But it will also redeploy on push to any branch whatsoever that changes `aws/devnet/**`, which I don't think is what we want.

Similar comment applies to all other workflows in this PR.

DaMandal0rian replied:
It will only run a plan, the apply/destroy will only run if input is set to yes.
https://github.com/subspace/infra/pull/183/files#diff-31842ccef98d0aadb01a72552ac62f9015e5139c67ebc0a9bc7d65ef272a1850R25-R32

So lets say i create a new branch and push to remote, it will show me the plan.

nazar-pc commented on .github/workflows/devnet_main_deploy.yml:
```diff
@@ -0,0 +1,27 @@
+
+jobs:
+  deploy:
+    uses: ./.github/workflows/terraform_template_deploy.yml
```
Why `./.github` and not just `.github`?

DaMandal0rian replied:
It will complain in IDE that it can't find the re-usable workflow

nazar-pc commented on .github/workflows/terraform_gh_runner.yml:
```diff
@@ -1,16 +1,18 @@
     paths:
-      - './github-runners/terraform/base/**'
-  workflow_dispatch:
+      - "./github-runners/terraform/base/**"
```
Similarly here `./` seems unnecessary

nazar-pc commented on .github/workflows/terraform_gh_runner.yml:
```diff
@@ -1,16 +1,18 @@
 jobs:
   terraform_gh_runner:
     runs-on: ubuntu-latest
+    env:
+      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
It doesn't seem to be used, why setting?

DaMandal0rian replied:
The env variable needs to be set when calling github api, otherwise an error is thrown.

nazar-pc replied:
What and where calls those APIs? I don't see any usage of it.

DaMandal0rian replied:
The GITHUB_TOKEN is needed in CI or bash cli for API calls. The API is called with curl via https://github.com/subspace/infra/pull/183/files/99a5ac25712fdda531fd88026a12f96aafeacb83#diff-b46b68b6df852ad5f8fc96162f55c2fe198f6d53eb2fe7d7ce12fcd4b2650ba6R39-R43

> If you want to use the API in a GitHub Actions workflow, GitHub recommends that you authenticate with the built-in GITHUB_TOKEN

nazar-pc replied:
What you linked doesn't use `GITHUB_TOKEN` environment variable, it uses `$token` that is obtained in a different way.

DaMandal0rian replied:
Ok, for $token in the auth header of the curl request that is the pat token to authenticate to the repo and get the runner token. If i remove GH_TOKEN you will see an error in ci that you need to have GH_TOKEN to call github apis from workflow :). I wasn't using it before, but added it to fix the problem. Here is an old version:
https://github.com/subspace/infra/blob/bd296590f29cdd730ab29869c7ade1258cb395ae/.github/workflows/terraform_gh_runner.yml

nazar-pc replied:
Then show me where it is used :man_shrugging: You have linked to `curl` request that doesn't use this environment variable as a reason for this environment variable to exist. It makes no sense to me.

DaMandal0rian replied:
@nazar-pc removed it, here you go:
https://github.com/subspace/infra/actions/runs/6484473796/job/17608384650#step:5:24

nazar-pc replied:
Okay, so it is not for `curl`, it was for `gh` CLI command, now it makes sense.

With that knowledge we understand why it is necessary. I think what we should do now in a secure way without using heavy tools like setting repo (I'm fairly certain simply setting environment variable wasn't) without using heavy tools like setting repo's secrets from workflow is to use outputs and mask them such that they are not visible in logs: https://github.com/orgs/community/discussions/25225#discussioncomment-3246942

That is the goal here: to pass the token from one step into another. What we had in earlier versions of this PR are various suboptimal/incorrect ways of achieving that ultimate goal.

DaMandal0rian replied:
correct, github cli, I should have been more clear about that "The GITHUB_TOKEN is needed in CI or bash cli for API calls" but couldn't remember exactly since it was a while ago. That being said the solution of masking the secret and just passing it is nice if using the ephemeral runners maybe but this solution and workflow is for the dedicated runners, where i need to retain the secret so I can unregister and delete runner if need be and remove it from github. See https://docs.github.com/en/free-pro-team@latest/rest/actions/self-hosted-runners?apiVersion=2022-11-28#delete-a-self-hosted-runner-from-an-organization

nazar-pc replied:
But you mentioned that tokens only live for 1 hour, so you'll have to retrieve fresh token anyway. Why retaining it then at all?

DaMandal0rian replied:
The token is not needed to deregister, a force delete can be done. So i've used the masking technique.

nazar-pc commented on .github/workflows/terraform_gh_runner.yml:
```diff
@@ -19,40 +21,47 @@ jobs:
 
       - name: Run Bash Script
+        env:
+          GH_TOKEN: ${{ github.token }}
```
You're using `secrets.GITHUB_TOKEN` above and `github.token` here, which are both the same thing. Would be nice to make it consistent.

But this environment doesn't seem to be used either.

DaMandal0rian replied:
It is used with github api, will make consistent. It is needed for api calls to get the PAT token for github runner registration.

nazar-pc replied:
GitHub API again? Which API, I don't see any usage of it.

DaMandal0rian replied:
https://github.com/subspace/infra/pull/183/files/99a5ac25712fdda531fd88026a12f96aafeacb83#diff-b46b68b6df852ad5f8fc96162f55c2fe198f6d53eb2fe7d7ce12fcd4b2650ba6R39-R43

nazar-pc replied:
How is linked code related to `GH_TOKEN`?

DaMandal0rian replied:
Ok, for $token in the auth header of the curl request that is the pat token to authenticate to the repo and get the runner token. If i remove GH_TOKEN you will see an error in ci that you need to have GH_TOKEN to call github apis from workflow :). I wasn't using it before, but added it to fix the problem. Here is an old version:
https://github.com/subspace/infra/blob/bd296590f29cdd730ab29869c7ade1258cb395ae/.github/workflows/terraform_gh_runner.yml

nazar-pc commented on .github/workflows/terraform_gh_runner.yml:
```diff
@@ -19,40 +21,47 @@ jobs:
-            token=${{ secrets.PAT_TOKEN }}
+          # GitHub repository and access token
+          repo="subspace/infra"
+          token=${{ secrets.PAT_TOKEN }}
```
A third token, can you explain what this third token is and why do we need it? I would have expected that workflow should be able to use `secrets.GITHUB_TOKEN`/`github.token` instead.

DaMandal0rian replied:
PAT token to interact with repo for creating a token for github runner registration. The `secrets.GITHUB_TOKEN` token can't be used for that purpose.

nazar-pc replied:
So you're using personal token for that? Is that a recommended approach? Seems fragile and generally problematic to me.

DaMandal0rian replied:
Yes, it's the recommended approach https://docs.github.com/en/rest/guides/getting-started-with-the-rest-api?apiVersion=2022-11-28&tool=curl#authenticating

GH_TOKEN does not have enough permissions on the repo/org level.

nazar-pc commented on .github/workflows/terraform_gh_runner.yml:
```diff
@@ -19,40 +21,47 @@ jobs:
-            # Store the token as a secret in GitHub Actions
-            gh secret set RUNNER_TOKEN -r "$repo" -b "$runner_token"
+          # Store the token as a secret in GitHub Actions
+          gh secret set RUNNER_TOKEN -r "$repo" -b "$runner_token"
```
This makes little sense, why are we setting it as environment variable in two different ways AND secrets? We only seem to use the first environment variable in the next step.

All this token management looks like a complete mess to me. A bunch of random things, some of which are used and some are not, but not clear why.

Comments were supposed to help, but these are exactly the kinds of comments that shouldn't be written in the first place: they say exactly the same thing that code does, so I essentially read the same information twice, which isn't helpful. Comments should add useful information, explaining why things are done, how they are useful but none of this is currently present.

DaMandal0rian replied:
There are essentially 3 tokens, `secrets.GITHUB_TOKEN` for using github api. `PAT_TOKEN` to access repository and create a runner registration token `runner_token` to register a runner for github actions. Yeah removed the extra step of setting it again since stored in  $GITHUB_ENV, was running as a bash script before and forgot to remove that part.

nazar-pc replied:
But why do you store it in the environment variable at all? It is already stored in secrets. Or why is it stored in secrets if it is only necessary in environment variable for the next step. Also storing it in enviromnent variable probably means it will be printed in CI run in logs, which is a security issue.

This is why it looks like a mess.

DaMandal0rian replied:
Storing as the env variable in that step will not print it to CI, its just echoing the string into GITHUB_ENV. However I have removed it and just used the one from the secret at https://github.com/subspace/infra/pull/183/files#diff-b46b68b6df852ad5f8fc96162f55c2fe198f6d53eb2fe7d7ce12fcd4b2650ba6R63

nazar-pc replied:
Why still doing `export RUNNER_TOKEN="$runner_token"`?

What about race conditions, it is possible that there are multiple tokens issued concurrently in different jobs. I'm also wondering if there is a limit for number of these tokens or whether there is an expiration limit for them. Because if there is a limit we need to reuse them, if there is no expiration we can probably just store them in secrets once and never re-request.

DaMandal0rian replied:
That was removed earlier the export statement, shouldn't be there if you check the file. PAT tokens can set an expiry date. Tokens that have expired can be regenerated, giving you a duplicate token with the same properties as the original. It's good practice to rotate though.

nazar-pc replied:
I am asking about `$runner_token` that this workflow stores in secrets, not PAT.

DaMandal0rian replied:
The token expires after one hour.
https://docs.github.com/en/rest/actions/self-hosted-runners?apiVersion=2022-11-28#create-a-registration-token-for-an-organization

DaMandal0rian committed:
minor fixes

- make tokens consistent and adjust some paths.
- add comments
- remove extra env variable set

DaMandal0rian committed:
just use secret for token and not env

nazar-pc reviewed (commented):
CI failed with `Error: Invalid command-line option`, which I think means something needs to be fixed.

DaMandal0rian committed:
Remove terrform init backend config

- this is not needed with terraform cloud
- test (remove GH_TOKEN)

DaMandal0rian committed:
add token masking

DaMandal0rian committed:
remove tfvars secret and add branch input

- remove tfvars secret and storing in github, use the encrypted terraform.tfvars file with transcrypt.
- add branch as input
- change API token name and remove vars secret file
- change path for dedicated runners

DaMandal0rian committed:
update to gemini-3g

DaMandal0rian committed:
add more inputs

- add genesis hash input
- add terraform apply and destroy inputs

vedhavyas reviewed (commented):
I maybe be missing something here but looks like we can only run one ephemeral devnet at a time since TF uses the same working directory. This would also be a problem if multiple people want to deploy at the same time.

If its aws, can we deploy new instances instead ?
As for the hetzner, we should not allow new deployments unless the workspace is already destroyed so that machine can be re-used.

Though I'm not sure if this is out of scope for this PR.
